plugins/pixart-rf/fu-pxi-device.c

```diff
@@ -0,0 +1,611 @@
+		return FALSE;
+
+	/* prepare write fw into device */
+	chunks = fu_chunk_array_new_from_bytes (fw, 0x0, 0x0, FU_PXI_DEVICE_OBJECT_SIZE_MAX);
+	if (!fu_pxi_device_check_support_resume (self, firmware, &error_local)) {
+		g_debug ("do not resume: %s", error_local->message);
+		self->offset = 0;
+		self->checksum = 0;
+	}
+
+	/* write fw into device */
+	fu_device_set_status (device, FWUPD_STATUS_DEVICE_WRITE);
+	for (guint i = self->offset; i < chunks->len; i++) {
+		FuChunk *chk = g_ptr_array_index (chunks, i);
+		if (!fu_pxi_device_write_chunk (self, chk, error))
+			return FALSE;
+		fu_device_set_progress_full (device, (gsize) i, (gsize) chunks->len);
+	}
+
+	/* fw upgrade command */
+	if (!fu_pxi_device_fw_upgrade (self, firmware, error))
+		return FALSE;
+
+	/* send device reset command */
+	return fu_pxi_device_reset (self, error);
+}
+
+static gboolean
+fu_pxi_device_fw_get_info (FuPxiDevice *self, GError **error)
+{
+	guint8 res[FU_PXI_DEVICE_OTA_BUF_SZ] = { 0x0 };
+	guint8 opcode = 0x0;
+	guint16 checksum = 0;
+	g_autofree gchar *version_str = NULL;
+	g_autoptr(GByteArray) req = g_byte_array_new ();
+
+	fu_byte_array_append_uint8 (req, PXI_HID_DEV_OTA_OUTPUT_REPORT_ID);
+	fu_byte_array_append_uint8 (req, FU_PXI_DEVICE_CMD_FW_GET_INFO);
+	if (!fu_udev_device_pwrite_full (FU_UDEV_DEVICE (self), 0,
+					 req->data, req->len, error))
+		return FALSE;
+
+	res[0] = PXI_HID_DEV_OTA_FEATURE_REPORT_ID;
+	res[1] = FU_PXI_DEVICE_CMD_FW_GET_INFO;
+	if (!fu_pxi_device_get_feature (self, res, FU_PXI_DEVICE_FW_INFO_RET_LEN + 1, error))
+		return FALSE;
+
+	if (g_getenv ("FWUPD_PIXART_RF_VERBOSE") != NULL)
+		fu_common_dump_raw (G_LOG_DOMAIN, "req", (guint8 *) req, sizeof(req));
+	if (!fu_common_read_uint8_safe (res, sizeof(res), 0x2, &opcode, error))
+		return FALSE;
+	if (opcode != FU_PXI_DEVICE_CMD_FW_GET_INFO) {
+		g_set_error (error,
+			     G_IO_ERROR,
+			     G_IO_ERROR_FAILED,
+			     "FwGetInfo opcode invalid 0x%02x",
+			     opcode);
+		return FALSE;
+	}
+	/* set current version */
+	version_str = g_strndup ((gchar *) res + 0x4, 5);
+	fu_device_set_version (FU_DEVICE (self), version_str);
+
+	/* add current checksum */
+	if (!fu_common_read_uint16_safe (res, sizeof(res), 0x9,
+					 &checksum, G_LITTLE_ENDIAN, error))
+		return FALSE;
+
+	/* success */
+	return TRUE;
+}
+
+static gboolean
+fu_pxi_device_probe (FuDevice *device, GError **error)
+{
+	/* set the physical ID */
+	return fu_udev_device_set_physical_id (FU_UDEV_DEVICE (device), "hid", error);
+}
+
+static gboolean
+fu_pxi_device_setup (FuDevice *device, GError **error)
+{
+	FuPxiDevice *self = FU_PXI_DEVICE (device);
+	if (!fu_pxi_device_fw_ota_init (self, error))
+		return FALSE;
+	if (!fu_pxi_device_fw_get_info (self, error))
+		return FALSE;
+	return TRUE;
+}
+
+static void
+fu_pxi_device_init (FuPxiDevice *self)
+{
+	fu_device_add_flag (FU_DEVICE (self), FWUPD_DEVICE_FLAG_UPDATABLE);
+	fu_device_add_flag (FU_DEVICE (self), FWUPD_DEVICE_FLAG_NO_GUID_MATCHING);
```

Comment:
Is this flag really needed?  It seems like this is a straightforward USB device.  Am I wrong?

Reply:
> Am I wrong?

Of course not! I've removed it :)